migrations/entitlements/migration_test.go

```diff
@@ -0,0 +1,2109 @@
+	ref = capValue.BorrowType.(*interpreter.ReferenceStaticType)
+	require.Equal(t,
+		interpreter.NewEntitlementSetAuthorization(
+			inter,
+			func() []common.TypeID { return []common.TypeID{"A.0000000000000001.C.E"} },
+			1,
+			sema.Conjunction,
+		),
+		ref.Authorization,
+	)
+}
+
+func TestMigrateDictOfWithTypeValueKey(t *testing.T) {
+	t.Parallel()
+
+	address1 := [8]byte{0, 0, 0, 0, 0, 0, 0, 1}
+	address2 := [8]byte{0, 0, 0, 0, 0, 0, 0, 2}
+
+	storage := NewTestLedger(nil, nil)
+	rt := NewTestInterpreterRuntime()
+
+	accountCodes := map[common.Location][]byte{}
+	interfaces := map[common.Location]*TestRuntimeInterface{}
+
+	runtimeInterface1 := &TestRuntimeInterface{
+		Storage: storage,
+		OnEmitEvent: func(event cadence.Event) error {
+			return nil
+		},
+		OnGetSigningAccounts: func() ([]runtime.Address, error) {
+			return []runtime.Address{address1}, nil
+		},
+		OnGetCode: func(location common.Location) (bytes []byte, err error) {
+			return accountCodes[location], nil
+		},
+		OnResolveLocation: MultipleIdentifierLocationResolver,
+		OnGetAccountContractCode: func(location common.AddressLocation) (code []byte, err error) {
+			code = accountCodes[location]
+			return code, nil
+		},
+	}
+	runtimeInterface1.OnUpdateAccountContractCode = func(location common.AddressLocation, code []byte) error {
+		accountCodes[location] = code
+		interfaces[location] = runtimeInterface1
+		return nil
+	}
+
+	runtimeInterface2 := &TestRuntimeInterface{
+		Storage: storage,
+		OnEmitEvent: func(event cadence.Event) error {
+			return nil
+		},
+		OnGetCode: func(location common.Location) (bytes []byte, err error) {
+			return accountCodes[location], nil
+		},
+		OnGetSigningAccounts: func() ([]runtime.Address, error) {
+			return []runtime.Address{address2}, nil
+		},
+		OnResolveLocation: MultipleIdentifierLocationResolver,
+		OnGetAccountContractCode: func(location common.AddressLocation) (code []byte, err error) {
+			code = accountCodes[location]
+			return code, nil
+		},
+	}
+	runtimeInterface2.OnUpdateAccountContractCode = func(location common.AddressLocation, code []byte) error {
+		accountCodes[location] = code
+		interfaces[location] = runtimeInterface2
+		return nil
+	}
+
+	nextTransactionLocation := NewTransactionLocationGenerator()
+
+	oldContract := []byte(`
+		access(all) contract C {
+			access(all) resource R {
+				access(all) fun foo() {}
+			}
+			access(all) fun makeR(): @R {
+				return <- create R()
+			}
+		}
+	`)
+
+	contract := []byte(`
+		access(all) contract C {
+			access(all) entitlement E
+			access(all) resource R {
+				access(E) fun foo() {}
+			}
+			access(all) fun makeR(): @R {
+				return <- create R()
+			}
+		}
+	`)
+
+	saveValues := []byte(`
+		import C from 0x1	
+
+		transaction {
+			prepare(signer: auth(Storage, Capabilities) &Account) {
+				let r1 <- C.makeR()
+				let r2 <- C.makeR()
+				let rType = ReferenceType(entitlements: [], type: r1.getType())!
```

Comment:
I'm not sure whether this test exercises the issue we are trying to cover. The values should be stored using the 'older' version of cadence, but here it stores using Cadence 1.0 (`ReferenceType(entitlements: [], type: r1.getType())!` is the new format of reference with the entitlements), right?

Reply:
Correct yeah. I don't think there's a way to create an old reference type within the new version of Cadence, so this exact case I think will need to be tested via the emulator.

Reply:
Makes sense. I have added a test in the FVM for this particular case, but would need this PR to be merged.